File: wasm/verify/border_test.go
package main

import (
	"reflect"
	"regexp"
	"sort"
	"strings"
	"testing"

	"github.com/rohanthewiz/grmob/htmlout"
)

// The set of node types whose user-agent border has to be written back to
// nothing.
//
// Go is the authority (borderResetTypes in htmlout/tag.go) and the runtime
// restates it, so the two are compared here rather than remembered — the same
// treatment the tag, input-type and generic-tag tables get. A drift is silent
// in the worst way: the exported document and the live app draw the same node
// with and without a rule around it, and nothing errors on either side.
//
// Parsed with a regexp rather than through parseRuntimeTable, which reads
// key/value object literals; this one is a flat array of strings.
func TestRuntimeBorderResetTypesMatchGo(t *testing.T) {
	src := runtimeSource(t)

	m := regexp.MustCompile(`const BORDER_RESET_TYPES = new Set\(\[([^\]]*)\]\);`).FindStringSubmatch(src)
	if m == nil {
		t.Fatalf("grmob-runtime.js: no `const BORDER_RESET_TYPES = new Set([...])` found — if it " +
			"was renamed or spread over several lines, update this test rather than deleting it")
	}
	var got []string
	for _, q := range regexp.MustCompile(`"([^"]*)"`).FindAllStringSubmatch(m[1], -1) {
		got = append(got, q[1])
	}
	sort.Strings(got)

	want := htmlout.BorderResetTypes()
	if !reflect.DeepEqual(got, want) {
		t.Errorf("BORDER_RESET_TYPES is %v; htmlout.BorderResetTypes() is %v — a node type the "+
			"browser's own border survives on one web target and not the other", got, want)
	}
}

// Every member of the set has to be a node type the tag table knows, and the
// three <input> types that are *not* in it have to stay out.
//
// The set was keyed by tag while <button> was its only member and is keyed by
// node type now that text fields have joined, so the spellings changed from
// "button" to "Button" — a difference no compiler notices, since both sides
// are map keys of type string. A stale lowercase entry would simply never
// match, and the symptom would be the divergence this whole mechanism exists
// to close, showing up as a missing rule rather than as a failure.
func TestBorderResetTypesAreNodeTypes(t *testing.T) {
	tags := htmlout.Tags()
	for _, nodeType := range htmlout.BorderResetTypes() {
		if _, ok := tags[nodeType]; !ok {
			t.Errorf("borderResetTypes has %q, which is not a node type in htmlout.Tags() — "+
				"the set is keyed by core node type, not by HTML tag", nodeType)
		}
	}
	// The controls the user agent draws in their entirety. Resetting their
	// border would be asking the browser not to draw the box that *is* the
	// checkbox; a range track has no border to begin with. See
	// borderResetTypes for why keying by tag would have swept both in.
	for _, nodeType := range []string{"Checkbox", "Slider"} {
		if htmlout.ResetsUABorder(nodeType) {
			t.Errorf("%s is in borderResetTypes — the user agent draws that control itself, "+
				"and the Go style is not meant to own its frame", nodeType)
		}
	}
}

// The set is only half the contract; the other half is that the runtime
// actually consults it, and consults it in the *negative* arm.
//
// Putting the reset in the positive arm, or writing it as a separate guarded
// assignment, would both look right and break styleFromGrMob's totality rule:
// every property it manages is assigned on every call, so that a field back at
// its zero value clears the declaration an earlier patch left behind. A
// guarded `if (BORDER_RESET_TYPES.has(...)) out.border = "none"` would leave a
// button that loses its border showing the last one it had.
func TestRuntimeResetsTheUserAgentBorder(t *testing.T) {
	src := runtimeSource(t)
	const expr = `: (BORDER_RESET_TYPES.has(nodeType) ? "none" : "");`
	if !strings.Contains(src, expr) {
		t.Errorf("grmob-runtime.js: styleFromGrMob is missing %q — without it a <button> with no "+
			"border in its style keeps the browser's, which is the divergence "+
			"components.Button's EmphasisGhost exposed", expr)
	}
}

File: wasm/verify/border_test.mjs
// The user-agent border reset, against the minimal DOM.
//
// htmlout writes the same declaration into a static document and is tested
// there; this is the live half, where the interesting property is the one a
// static export cannot have — that a button which *loses* its border falls
// back to "none" rather than to the browser's own rule.
//
// styleFromGrMob is total: every property it manages is assigned on every
// call, because an update-style patch carries the whole new Style and a field
// back at its zero value means "unset now". For every other property "unset"
// is the empty string, which drops the inline declaration and lets the
// cascade decide. For a <button>'s border, letting the cascade decide is the
// bug — the user-agent stylesheet is what draws the 2px outset rule
// comps.Button's EmphasisGhost was documented as not having. So this one
// property has three values rather than two, and the third is what these
// tests hold.

import test from "node:test";
import assert from "node:assert/strict";

import { loadRuntime, nodeAt } from "./load.mjs";

// One tree, one element under test, plus the handle to patch its style. Same
// shape as a11y_test.mjs's.
function mount(children) {
    const rt = loadRuntime();
    rt.GrMob.mount(
        JSON.stringify({ Type: "Column", Children: children })
    );
    rt.drainFrames();
    return { rt, at: (i) => nodeAt(rt.document, `root/${i}`) };
}

const button = (style) => ({ Type: "Button", Props: { label: "Skip" }, Style: style });

test("a button with no border is talked out of the browser's", () => {
    // The EmphasisGhost case: transparent fill, colored label, nothing said
    // about a border. Both natives draw none; without this the web drew the
    // user agent's, and no core.BorderWidth(0) could remove it because
    // emitting nothing is exactly what leaves the browser in charge.
    const { at } = mount([button({ Background: "#00000000", TextColor: "#007AFF" })]);
    assert.equal(at(0).style.border, "none");
});

test("a button with a border draws the one it was given", () => {
    // The EmphasisOutlined case. The reset must not swallow a real border.
    const { at } = mount([button({ BorderWidth: 1, BorderColor: "#007AFF" })]);
    assert.equal(at(0).style.border, "1px solid #007AFF");
});

test("half a border is no border, and still resets", () => {
    // Both halves are required on all four targets — Compose skips its
    // Modifier.border unless width and color are both set, and SwiftUI's
    // grMobBorder guards the same way — so a width with no color must reach
    // the reset rather than emitting an incomplete declaration.
    const { at } = mount([
        button({ BorderWidth: 1 }),
        button({ BorderColor: "#007AFF" }),
    ]);
    assert.equal(at(0).style.border, "none");
    assert.equal(at(1).style.border, "none");
});

test("a div with no border says nothing about one", () => {
    // The reset is scoped to the node types the browser draws a border on. A
    // <div> has none, so "none" there would be a declaration that means
    // nothing and that an author's own stylesheet would then have to fight.
    const { at } = mount([{ Type: "Box", Style: { Background: "#fff" } }]);
    assert.equal(at(0).style.border, "");
});

test("a text field with no border is talked out of the browser's too", () => {
    // The three field types joined the set once both bundled themes gave
    // Components.Input and Components.TextArea a frame of their own — before
    // that, resetting the browser's rule left the control undrawn. A field
    // whose style states no border is now borderless everywhere, which is
    // what it already was on both phones.
    const { at } = mount([
        { Type: "Input", Props: { value: "", placeholder: "Search" }, Style: { Background: "#fff" } },
        { Type: "TextArea", Props: { value: "" }, Style: { Background: "#fff" } },
        { Type: "InputPassword", Props: { value: "" }, Style: { Background: "#fff" } },
    ]);
    assert.equal(at(0).style.border, "none");
    assert.equal(at(1).style.border, "none");
    assert.equal(at(2).style.border, "none");
});

test("a themed field draws the frame the theme gave it", () => {
    // The ordinary case, and the one the reset must not swallow: core.Input
    // carries Components.Input, which both bundled themes now fill in.
    const { at } = mount([{
        Type: "Input",
        Props: { value: "", placeholder: "Name" },
        Style: { Background: "#FFFFFF", BorderColor: "#8E8E93", BorderWidth: 1 },
    }]);
    assert.equal(at(0).style.border, "1px solid #8E8E93");
});

test("the user agent keeps its checkbox and its slider", () => {
    // The case that made the set node-type-keyed rather than tag-keyed. All
    // five of these are <input> elements; on a checkbox and a range the
    // browser draws the whole control, so its border is the box itself and
    // not chrome the Go style is meant to own.
    // Both carry a Style, because a node with none never reaches
    // styleFromGrMob at all and would pass this by never being asked.
    const { at } = mount([
        { Type: "Checkbox", Props: { checked: false }, Style: { Background: "#fff" } },
        { Type: "Slider", Props: { value: 0.5 }, Style: { Background: "#fff" } },
    ]);
    assert.equal(at(0).style.border, "");
    assert.equal(at(1).style.border, "");
});

test("a button that loses its border does not get the browser's back", () => {
    // The totality rule, and the reason the reset lives in the false arm of
    // the same expression rather than in a guard of its own. A guarded write
    // would leave the old declaration standing; an unguarded "" would hand the
    // element back to the user-agent stylesheet. Neither is what an outlined
    // button turning ghost is asking for.
    const { rt, at } = mount([button({ BorderWidth: 1, BorderColor: "#007AFF" })]);
    assert.equal(at(0).style.border, "1px solid #007AFF");

    rt.GrMob.patch(JSON.stringify([{
        Type: "update-style",
        TargetID: "root/0",
        Changes: { Background: "#00000000" },
    }]));
    rt.drainFrames();

    assert.equal(at(0).style.border, "none");
});
